// jovo-core/dist/src/TestHost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Log_1 = require("./Log");
class TestHost {
    constructor(req) {
        this.hasWriteFileAccess = true; // tests shouldn't write to filesystem
        this.failed = false;
        this.headers = { 'jovo-test': 'TestHost' };
        this.req = req;
        this.$request = req;
    }
    /**
     * Full request object
     * @returns {any}
     */
    getRequestObject() {
        return this.$request;
    }
    /**
     * Save the response
     * @param obj
     * @returns {Promise<any>}
     */
    setResponse(obj) {
        return new Promise((resolve) => {
            this.res = obj;
            resolve();
        });
    }
    /**
     * Return the previously set response object
     */
    getResponse() {
        return this.res;
    }
    /**
     * Save the error and set failed flag
     * @param error
     */
    fail(error) {
        Log_1.Log.error('TestHost.fail: ');
        Log_1.Log.error(error);
        this.err = error;
        this.failed = true;
    }
    /**
     * Return the previously set error
     * @returns Error
     */
    getError() {
        return this.err;
    }
    /**
     * Returns true if an error occurred while handling the request
     */
    didFail() {
        return this.failed;
    }
}
exports.TestHost = TestHost;
